arch/x86/kvm/vmx/pkvm/hyp/pkvm.c

```diff
@@ -157,15 +157,10 @@ unsigned long __pkvm_teardown_shadow_vm(int shadow_vm_handle)
 
 	pkvm_shadow_ept_deinit(&vm->sept_desc);
 
-	pkvm_pgstate_pgt_deinit(vm);
-
-	list_for_each_entry_safe(ptdev, tmp, &vm->ptdev_head, vm_node) {
-		pkvm_spin_lock(&vm->lock);
-		list_del(&ptdev->vm_node);
-		pkvm_spin_unlock(&vm->lock);
+	list_for_each_entry_safe(ptdev, tmp, &vm->ptdev_head, vm_node)
+		pkvm_detach_ptdev(ptdev, vm);
 
-		pkvm_detach_ptdev(ptdev);
-	}
+	pkvm_pgstate_pgt_deinit(vm);
```

Comment:
Again, just a note, not a suggestion: I understand you moved `pkvm_pgstate_pgt_deinit()` because `pkvm_pgstate_pgt_update_coherency()` should not be called for an already deinitialized pgstate_pgt, and it's probably ok per se, but just FYI it conflicts with my implementation of shadow iommu page tables for legacy mode as I explained in https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/4233618/4/arch/x86/kvm/vmx/pkvm/hyp/iommu.c#381. So I will need to rework it somehow.

Reply:
I see. So in legacy mode, detaching will sync ptdev->vpgt to ptdev->pgt, but because the mapping in vpgt is not undonated yet, the sync in legacy mode will be failed. But seems this failure is harmless because later host VMM will attach these devices to another IOMMU domain which will trigger IOMMU to sync again. So probably just need to make the ptdev->pgt sync stopped once there is undonated page?

Reply:
Yes, moving deinit before detaching (or maybe some other way to ensure the undonate is done before the iommu sync during detach) seems easier and cleaner to me.

Reply:
Deinit has been moved before detaching in the new version of the pull request